### PR TITLE
Insert process instance keys into `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` ColumnFamily

### LINKED ISSUE #13341: Insert into the `PROCESS_INSTANCE_KEYS_BY_DEFINITION_KEY` ColumnFamily
## Description
In #13340 we have created this ColumnFamily. We now need to fill it with data.

- Add a method to insert data into this ColumnFamily
    - Include tests!
- Use this method to insert data when starting a new process instance
    - Also when starting a process anywhere
- Use this method to insert data upon calling a CallActivity

**Blocked by**
#13340

### LINKED ISSUE #13342: Remove from `PROCESS_INSTANCE_KEYS_BY_DEFINITION_KEY` ColumnFamily
## Description
When a process has reached an end state we must remove it from this ColumnFamily

- Add a method to remove data from this ColumnFamily
    - Include tests!
- When a process is completed use this method to remove it from the ColumnFamily
- When a process is terminated use this method to remove it from the ColumnFamily
- When a call activity is completed / terminated use this method to remove it from the ColumnFamily

Please also change the `ElementInstanceStateTest#shouldNotLeakMemoryOnRemoval`. We must change the processInstanceRecord that's generated here to be of element type PROCESS.

**Blocked by**
#13340

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

This PR makes it so we can insert data into the ColumnFamily. The ColumnFamily keeps track of the process instances that are running for a specific process definition key. This will allow us to later determine which process instances we need to terminate when deleting a process definition.

## Related issues

closes #13341 
closes #13342 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
Can be reviewed after #13371 is merged

**Comment by remcowesterhoud** [comment]
@korthout I am aware the CI is failing. For it to succeed I need to merge #13376 into this one. Thing is, that one is based on this PR as it requires it for writing decent unit tests.

I guess it makes most sense to review this one, approve it but not merge it yet (cause obviously the code is perfect as-is 🙃). Then we can review the other one, merge it into this one and merge it into main as a batch.

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java:200]: 💭 If we make `processDefinitionKey` a foreign key to the process definition, we'd be able to completely avoid the situation that we saw in:
- https://github.com/camunda/zeebe/issues/13254

I previously [wrote about that](https://github.com/camunda/zeebe/issues/13254#issuecomment-1617550522):
>If we had such a check, the corrupted state would've been avoided. The affected partition would've been DEAD but recoverable after a patch. Such a check would avoid all errors (including the unknown ones) related to process instance creation where the process definition is actually not stored in state.
>
>Keeping track of this foreign key relation would come at the cost of performance (two additional IO operations on process instance creation: one to store the additional entry in a new column family, and one to check the foreign key relation when the instance is being created in the state).

But, if we have this in place anyways to support #9576, then that cost is significantly lower. Only the foreign key assertion (lookup) on insert would be needed. Let's consider this.

@remcowesterhoud If you agree, we could move this to a separate issue. That allows us to benchmark the difference to this change.

🔧 We should benchmark the current change anyways as well.

**Review by korthout — APPROVED** [review]
Nice work @remcowesterhoud 👏 

👍 Clear changes

🔧 Please make sure to benchmark this before merging, as an additional insert is performed for every created element instance.

LGTM 👍

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java:200]: > If we make processDefinitionKey a foreign key to the process definition, we'd be able to completely avoid the situation that we saw in:

I get what you're saying, but this also means our `ProcessState` and `ElementInstanceState` would be aware of each others existence. I'm not quite sure how I feel about that.

**Comment by remcowesterhoud** [comment]
> Please make sure to benchmark this before merging, as an additional insert is performed for every created element instance.

I will merge #13376 into this one before running the benchmark. I do want to clarify that we don't do an extra insert for every element instance. We only do the insert if the element type is `PROCESS`. So we'd only do the insert once per process instance.

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java:207]: That's a fair point, but the [timers have a foreign key to element instances](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbTimerInstanceState.java#L45-L49), and incidents have a foreign key [to element instances](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbIncidentState.java#L57) and [to jobs](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbIncidentState.java#L65). And similarly, we can add a relation from process instance (element instance of type process) to process definition. I think this is reasonable for relational data constraints.

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java:207]: I wasn't aware we already had such dependencies. In that case I don't see why not 🤷 Would you mind creating the issue? I think you have more context on the original issue.

**Pull request closed without merging**
